File: PIAPI/PIForm.m
//
//  PIForm.m
//  PIAPI
//
//  Created by Étienne VALLETTE d'OSIA on 02/12/2013.
//  Copyright (c) 2013 Zengularity. All rights reserved.
//

#import "PIForm.h"

@implementation PIField

@synthesize name = _name;
@synthesize type = _type;
@synthesize defaultValue = _defaultValue;
@synthesize multiple = _multiple;

+ (PIField *)FieldWithJson:(NSDictionary *)jsonObject name:(NSString *)name
{
    return [[PIField alloc] initWithJson:jsonObject name:name];
}

- (PIField *)initWithJson:(NSDictionary *)jsonObject name:(NSString *)name
{
    self = [self init];
    _name = name;
    _type = jsonObject[@"type"];
    _defaultValue = jsonObject[@"default"];
    NSNumber *multiple = jsonObject[@"multiple"];
    _multiple = multiple && multiple.boolValue;;
    return self;
}

@end

@implementation PIForm

@synthesize name = _name;
@synthesize method = _method;
@synthesize rel = _rel;
@synthesize enctype = _enctype;
@synthesize action = _action;
@synthesize fields = _fields;

+ (PIForm *)FormWithJson:(NSDictionary *)jsonObject name:(NSString *)name
{
    return [[PIForm alloc] initWithJson:jsonObject name:name];
}

- (PIForm *)initWithJson:(NSDictionary *)jsonObject name:(NSString *)name
{
    self = [self init];
    _name = name;
    _method = jsonObject[@"method"];
    _rel = jsonObject[@"rel"];
    _enctype = jsonObject[@"enctype"];
    _action = jsonObject[@"action"];
    NSDictionary *fieldsJson = jsonObject[@"fields"];
    NSMutableDictionary *fields = [[NSMutableDictionary alloc] init];
    for (NSString *fieldName in fieldsJson) {
        NSDictionary *fieldJson = fieldsJson[fieldName];
        PIField *field = [PIField FieldWithJson:fieldJson name:fieldName];
        fields[fieldName] = field;
    }
    _fields = fields;

    return self;
}

- (PIField *)fieldForName:(NSString *)name
{
    return [_fields valueForKey:name];
}

@end

File: PIAPI/PIForm.swift
//
//  Form.swift
//  PIAPI
//
//  Created by Erwan Loisant on 24/11/14.
//  Copyright (c) 2014 Zengularity. All rights reserved.
//

import Foundation


struct PIField {
    let type: String
    let multiple: Bool
    let defaultValue: String?

    init(type: String, multiple: Bool, defaultValue: String?) {
        self.type = type
        self.multiple = multiple
        self.defaultValue = defaultValue
    }
    
    static func parseJson(json: JSON) -> PIField {
        return PIField(type: json["type"].stringValue!, multiple: json["type"].boolValue, defaultValue: json["default"].stringValue)
    }
}


struct PIForm {
    let name: String?
    let method: String
    let rel: String?
    let enctype: String
    let action: String
    let fields: Dictionary<String, PIField>

    init(name: String?, method: String, rel: String?, enctype: String, action: String, fields: Dictionary<String, PIField>) {
        self.name = name
        self.method = method
        self.rel = rel
        self.enctype = enctype
        self.action = action
        self.fields = fields
    }
    
    func defaultData() -> Dictionary<String, Array<String>> {
        return fields.filter { (name: String, field: PIField) -> Bool in field.defaultValue != nil }
              .mapValues { [$0.defaultValue!] }
    }

    static func parseJson(json: JSON) -> PIForm {
        return PIForm(
            name: json["name"].stringValue?,
            method: json["method"].stringValue!,
            rel: json["rel"].stringValue?,
            enctype: json["enctype"].stringValue!,
            action: json["action"].stringValue!,
            fields: json["fields"].dictionaryValue?.mapValues{PIField.parseJson($0)} ?? Dictionary<String, PIField>()
        )
    }
    
}

